docs/adr/0002-project-locks.md

```diff
@@ -0,0 +1,169 @@
+
+ ## Goals & Non-Goals
+ 
+ ### Goals
+ - Alignment on what Atlantis should or shouldn't be locking
+ - Focus on small scope changes related only to locking
+ - Supports all workflow use-cases 
+	 - Terraform w/o workspaces
+	 - Terraform w/ workspaces
+	 - Terragrunt monorepo
+ 
+### Non-Goals
+- Making changes/improvements to other packages/sub-systems
+  - Moving plan data storage into backend state
+- Avoid massive refactoring if possible
+- Focus on a singular workflow
+
+## Previous attempts
+ There have been a couple of PRs submitted that have either been reverted, or grown stale:
+
+- add path to WorkingDir [#2180](https://github.com/runatlantis/atlantis/pull/2180)
+	- base32 encodes the path provided
+	- Uses the unique base32 string to additional clone the repo to unique directories to avoid plan file overlap per project + workspace
+- add non-default workspaces to workflow hooks [#2882](https://github.com/runatlantis/atlantis/pull/2882)
+    - reintroduces #2180 changes to working_dir
+    - run hooks on the default workspace
+    - also run hooks on every project found
+    - significant increase in execution time, especially for Terragrunt users relying on terragrunt-atlantis-config
+- reduce the amount of git clone operations [#2921](https://github.com/runatlantis/atlantis/pull/2921)
+    - started to try to reduce execution time on #2882 implementation
+    - attempts to utilize [TF_DATA_DIR](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_data_dir) for workspaces and remove the workspace for the lock. 
+    - Clones only at BaseRepo/PR# and moves workspaces from Atlantis clones to TF workspace managed
+
+This is not to suggest the revival of these PRs in their current state but to act as a reference for additional focused solutions. 
+
+## Solution: Clone once + TF_DATA_DIR
+
+Take PR #2921 and re-implement locks on the terraform DATA_DIR only and the entire base repo + PR for git clones. This will focus solely on git operations and Terraform Data Directories. 
+
+### Clone once
+
+We should be avoiding altogether re-cloning the repo unless absolutely necessary. We currently attempt to clone 3 times in a single command execution.
+
+1) Pre-Workflow Hooks
+2) <Import/Plan/Apply>CommandRunner
+	a) determintes projects to run plans on
+3) ProjectCommandRunner
+	a) clones before actual plan execution
+
+We should clone one initially for the entire repo/PR. This will empower multiple workspaces/plans alongside TF_DATA_DIR without needing to re-clone.
+
+Here are the cases that will trigger a clone:
+
+	1) Initial clone of repo
+ 	2) Force Reclone 
+  		a) Error when attempting fetch + pull to update existing local git repo
+    		b) File operation on a path that doesn't exist (data loss/accidental deletes)
+
+In all other situations, we should be utilizing Git for its delta compressions and performing less intensive fetch + pull operations.
```

Comment:
If you are using the merge checkout strategy, The time between 2) and 3) can be hours/days/weeks, depending on what other PRs are doing. 

So, before running 3), you will at least to verify that you are still up to date before proceeding.  You will have to merge (again) if upsteam has been modified and you are using the merge strategy.